htsget-search/src/htsget/vcf_search.rs

```diff
@@ -0,0 +1,407 @@
+  /// start of a BGZF block). If we read the records pointed by that coffset until we
+  /// reach a different coffset, we can find out where the current block ends.
+  /// Therefore this can be used to only add the required bytes in the query results.
+  /// If for some reason we can't read correctly the records we fall back
+  /// to adding the maximum BGZF block size.
+  fn bytes_range_end(&self, reader: &mut vcf::Reader<bgzf::Reader<File>>) -> u64 {
+    if self.uncompressed() == 0 {
+      // If the uncompressed part is exactly zero, we don't need the next block
+      return self.compressed();
+    }
+    get_next_block_position(*self, reader).unwrap_or(self.compressed() + Self::MAX_BLOCK_SIZE)
+  }
+
+  fn to_string(&self) -> String {
+    format!("{}/{}", self.compressed(), self.uncompressed())
+  }
+}
+
+fn get_next_block_position(
+  block_position: VirtualPosition,
+  reader: &mut vcf::Reader<bgzf::Reader<File>>,
+) -> Option<u64> {
+  reader.seek(block_position).ok()?.compressed();
+  let next_block_index = loop {
+    let bytes_read = reader.read_record(&mut String::new()).ok()?;
+    let actual_block_index = reader.virtual_position().compressed();
+    if bytes_read == 0 || actual_block_index > block_position.compressed() {
+      break actual_block_index;
+    }
+  };
+  Some(next_block_index)
+}
+
+pub(crate) struct VcfSearch<'a, S> {
+  storage: &'a S,
+}
+
+impl<'a, S> VcfSearch<'a, S>
+where
+  S: Storage + 'a,
+{
+  const MIN_SEQ_POSITION: i32 = 1; // 1-based
+  const MAX_SEQ_POSITION: i32 = 37450; // see https://github.com/zaeleus/noodles/issues/25#issuecomment-868871298
+
+  pub fn new(storage: &'a S) -> Self {
+    Self { storage }
+  }
+
+  pub fn search(&self, query: Query) -> Result<Response> {
+    let (vcf_key, tbi_key) = self.get_keys_from_id(query.id.as_str());
+
+    match query.class {
+      Class::Body => {
+        let tbi_path = self.storage.get(&tbi_key, GetOptions::default())?;
+        let vcf_index = tabix::read(tbi_path).map_err(|_| HtsGetError::io_error("Reading TBI"))?;
+
+        let byte_ranges = match query.reference_name.as_ref() {
+          None => self.get_byte_ranges_for_all_variants(&vcf_index, &vcf_key)?,
+          Some(reference_name) => self.get_byte_ranges_for_reference_name(
+            vcf_key.as_str(),
+            reference_name,
+            &vcf_index,
+            &query,
+          )?,
+        };
+        self.build_response(query, &vcf_key, byte_ranges)
+      }
+      Class::Header => {
+        let byte_ranges = self.get_byte_ranges_for_header(vcf_key.as_str())?;
+        self.build_response(query, &vcf_key, byte_ranges)
+      }
+    }
+  }
+
+  /// Generate a key for the storage object from an ID
+  /// This may involve a more complex transformation in the future,
+  /// or even require custom implementations depending on the organizational structure
+  /// For now there is a 1:1 mapping to the underlying files
```

Comment:
Thanks, makes sense for now, although it'll make sense to sync with @victorskl in the future since he's knee deep in DRS and Gen3 right now (close to our internal/organizational ID scheme).

For now 1-1 mapping is totally acceptable/fine.